### PR TITLE
Add Docker dev environment

### PR DESCRIPTION
**Pull request by inductiveload**
Based on the same PHP base image as WS-export.

Uses docker-compose for easy remapping of ports, etc.

See also T276925

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+FROM php:7.3-apache-stretch
+
+WORKDIR "/var/www/html"
+
+EXPOSE 8000
+
+# work around https://github.com/docker-library/openjdk/blob/0584b2804ed12dca7c5e264b5fc55fc07a3ac148/8-jre/slim/Dockerfile#L51-L54
+RUN mkdir -p /usr/share/man/man1
```
Comment (dayllanmaza) [inline Dockerfile:8]: I don't think you need this

**Review by dayllanmaza — CHANGES_REQUESTED** [review]
Looks good except for a couple tiny things.
I don't really know much about this project and it's dependencies but the container runs

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+IAUPLOAD_DOCKER_UID=1000
+IAUPLOAD_DOCKER_GID=100
+```
+and set the variable to match your system and the port you want.
+4. Register an oAuth consumer on [Meta](https://meta.wikimedia.org/wiki/Special:OAuthConsumerRegistration)
+   with a callback of `http://localhost:8000/oauth/callback`
+5. Edit `config.ini` to add your consumer key and secret
+6. Build the Docker image: `docker-compose build`
+7. Start the container: `docker-compose up -d`
+8. Install dependencies: `docker compose exec ia-upload composer install`
```
[the remark this answers is not in the document]
Comment (johnbeard) [reply CONTRIBUTING.md]: Good catch (also the 8. in the next line)

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+version: "3.9"
```
[the remark this answers is not in the document]
Comment (johnbeard) [reply docker-compose.yml]: ¯\_(ツ)_/¯ I haven no clue, I cargo-culted that from somewhere! 3.7 it is.

File: Dockerfile (hunk 1 above)
Comment (johnbeard) [reply Dockerfile]: OK

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (inductiveload) [reply CONTRIBUTING.md]: Good catch! Also the 8. in the line below.

File: docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (inductiveload) [reply docker-compose.yml]: I cargo-culted that from somewhere, I swear! 3.7 it is.

File: Dockerfile (hunk 1 above)
Comment (inductiveload) [reply Dockerfile]: OK

**Comment by inductiveload** [comment]
@samwilson gentle ping?

**Review by samwilson — COMMENTED** [review]
This looks great! Sorry for not reviewing sooner.

The only tiny change I'd make is to rename the `dev/` directory `docker/`, in line with our other repos.

**Comment by samwilson** [comment]
Actually, sorry, I think that's far too picky of me! hehe :) let's merge this.

**Comment by inductiveload** [comment]
> Actually, sorry, I think that's far too picky of me! hehe :) let's merge this.

Always feel free to make changes to any patches of mine if you want :-)

**Pull request merged**